From ab08c5418098ca142df0b91451d5c1d52b4d9b5d Mon Sep 17 00:00:00 2001
From: Michal Schmidt <mschmidt@redhat.com>
Date: Fri, 5 Sep 2014 19:49:51 -0400
Subject: [ethernet] bnx2x: fix possible panic under memory stress

Message-id: <1409946615-5862-47-git-send-email-mschmidt@redhat.com>
Patchwork-id: 91611
O-Subject: [RHEL7.1 PATCH 46/70] bnx2x: fix possible panic under memory stress
Bugzilla: 1089402
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1089402

Upstream commit:
commit 07b0f00964def8af9321cfd6c4a7e84f6362f728
Author: Eric Dumazet <edumazet@google.com>
Date:   Thu Jun 26 00:44:02 2014 -0700

    bnx2x: fix possible panic under memory stress

    While it is legal to kfree(NULL), it is not wise to use :
    put_page(virt_to_head_page(NULL))

     BUG: unable to handle kernel paging request at ffffeba400000000
     IP: [<ffffffffc01f5928>] virt_to_head_page+0x36/0x44 [bnx2x]

    Reported-by: Michel Lespinasse <walken@google.com>
    Signed-off-by: Eric Dumazet <edumazet@google.com>
    Cc: Ariel Elior <ariel.elior@qlogic.com>
    Fixes: d46d132cc021 ("bnx2x: use netdev_alloc_frag()")
    Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
index 50e9445..076acf1 100644
--- a/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
+++ b/drivers/net/ethernet/broadcom/bnx2x/bnx2x_cmn.c
@@ -803,7 +803,8 @@ static void bnx2x_tpa_stop(struct bnx2x *bp, struct bnx2x_fastpath *fp,
 
   return;
  }
- bnx2x_frag_free(fp, new_data);
+ if (new_data)
+  bnx2x_frag_free(fp, new_data);
 drop:
  /* drop the packet and keep the buffer in the bin */
  DP(NETIF_MSG_RX_STATUS,
-- 
1.7.1